--- haproxy-config.template.orig	2016-11-21 14:32:18.260057025 +0100
+++ haproxy-config.template	2016-11-21 15:23:23.172753955 +0100
@@ -89,6 +89,7 @@
 
 frontend public
   bind :{{env "ROUTER_SERVICE_HTTP_PORT" "80"}}
+  bind :::{{env "ROUTER_SERVICE_HTTP_PORT" "80"}} v6only
   mode http
   tcp-request inspect-delay 5s
   tcp-request content accept if HTTP
@@ -96,9 +97,14 @@
   # Remove port from Host header
   http-request replace-header Host (.*):.* \1
 
+  acl path_letsencrypt path_beg /.well-known/acme-challenge/
+  acl path_letsencrypt path_beg /.well-known/letsencrypt
+
   # check if we need to redirect/force using https.
   acl secure_redirect base,map_beg(/var/lib/haproxy/conf/os_edge_http_redirect.map) -m found
-  redirect scheme https if secure_redirect
+  redirect scheme https if secure_redirect !path_letsencrypt
+
+  use_backend be_http_%[env(LETSENCRYPT_HOST),map_beg(/var/lib/haproxy/conf/os_http_be.map)] if path_letsencrypt
 
   # Check if it is an edge route exposed insecurely.
   acl edge_http_expose base,map_beg(/var/lib/haproxy/conf/os_edge_http_expose.map) -m found
@@ -117,6 +123,7 @@
 # that terminates encryption in this router (edge)
 frontend public_ssl
   bind :{{env "ROUTER_SERVICE_HTTPS_PORT" "443"}}
+  bind :::{{env "ROUTER_SERVICE_HTTPS_PORT" "443"}} v6only
   tcp-request  inspect-delay 5s
   tcp-request content accept if { req_ssl_hello_type 1 }
 
